packages/pigeon/lib/cpp_generator.dart

```diff
@@ -0,0 +1,670 @@
+  return 'set$uppercased';
+}
+
+String? _cppTypeForBuiltinDartType(TypeDeclaration type) {
+  const Map<String, String> cppTypeForDartTypeMap = <String, String>{
+    'bool': 'bool',
+    'int': 'int64_t',
+    'String': 'std::string',
+    'double': 'double',
+    'Uint8List': 'std::vector<uint8_t>',
+    'Int32List': 'std::vector<int32_t>',
+    'Int64List': 'std::vector<int64_t>',
+    'Float64List': 'std::vector<double>',
+    'Map': 'flutter::EncodableMap',
+    'List': 'flutter::EncodableList',
+  };
+  if (cppTypeForDartTypeMap.containsKey(type.baseName)) {
+    return cppTypeForDartTypeMap[type.baseName];
+  } else {
+    return null;
+  }
+}
+
+String _cppTypeForDartType(TypeDeclaration type) {
+  return _cppTypeForBuiltinDartType(type) ?? type.baseName;
+}
+
+/// Generates the ".h" file for the AST represented by [root] to [sink] with the
+/// provided [options].
+void generateCppHeader(CppOptions options, Root root, StringSink sink) {
+  final Indent indent = Indent(sink);
+  if (options.copyrightHeader != null) {
+    addLines(indent, options.copyrightHeader!, linePrefix: '// ');
+  }
+  indent.writeln('// $generatedCodeWarning');
+  indent.writeln('// $seeAlsoWarning');
+  indent.addln('');
+  indent.writeln('#pragma once');
+  indent.writeln('#include <flutter/encodable_value.h>');
+  indent.writeln('#include <flutter/basic_message_channel.h>');
+  indent.writeln('#include <flutter/binary_messenger.h>');
+  indent.writeln('#include <flutter/standard_message_codec.h>');
+  indent.writeln('#include <map>');
+  indent.writeln('#include <string>');
+
+  indent.addln('');
+
+  if (options.namespace != null) {
+    indent.writeln('namespace ${options.namespace} {');
+  }
+
+  indent.addln('');
+  indent.writeln('/** Generated class from Pigeon. */');
+
+  for (final Enum anEnum in root.enums) {
+    indent.writeln('');
+    indent.write('enum class ${anEnum.name} ');
+    indent.scoped('{', '};', () {
+      int index = 0;
+      for (final String member in anEnum.members) {
+        indent.writeln(
+            '$member = $index${index == anEnum.members.length - 1 ? '' : ','}');
+        index++;
+      }
+    });
+  }
+
+  for (final Class klass in root.classes) {
+    indent.addln('');
+    indent.writeln(
+        '/** Generated class from Pigeon that represents data sent in messages. */');
+    indent.write('class ${klass.name} ');
+    indent.scoped('{', '};', () {
+      indent.scoped('public:', '', () {
+        for (final NamedType field in klass.fields) {
+          final HostDatatype hostDatatype = getHostDatatype(field, root.classes,
+              root.enums, (NamedType x) => _cppTypeForBuiltinDartType(x.type));
+          indent.writeln('${hostDatatype.datatype} ${_makeGetter(field)}();');
+          indent.writeln(
+              'void ${_makeSetter(field)}(${hostDatatype.datatype} setterArg);');
+          indent.addln('');
+        }
+        indent.write('flutter::EncodableMap ToEncodableMap(); ');
```

Comment:
We need to make these methods private.  We don't want people to rely on them or call them because we want to completely abstract away the wire format of pigeon.  I think the easiest thing is to make these private and make the serializer a `friend`.

Reply:
Done.  I think we might be able to make something a bit cleaner if we just put the encode and decode functions in the .cpp file and don't include them in the header, but this gets is what we want for now.